File: app/src/main/java/com/test/xyz/demo/domain/interactor/project/ProjectInteractorImpl.java
package com.test.xyz.demo.domain.interactor.project;

import com.google.common.base.Strings;
import com.test.xyz.demo.domain.model.github.GitHubRepo;
import com.test.xyz.demo.domain.repository.api.ProjectListRepository;

import java.util.List;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

class ProjectInteractorImpl implements ProjectInteractor {
    private static final String TAG = ProjectInteractorImpl.class.getName();

    private final ProjectListRepository projectListRepository;

    public ProjectInteractorImpl(ProjectListRepository projectListRepository) {
        this.projectListRepository = projectListRepository;
    }

    @Override
    public Observable<List<GitHubRepo>> getProjectList(String userName) {
        if (Strings.isNullOrEmpty(userName)) {
            return Observable.error(new IllegalArgumentException("Username must be provided!"));
        }

        return projectListRepository.getProjectList(userName)
                .subscribeOn(Schedulers.io());
    }

    @Override
    public Observable<GitHubRepo> getProjectDetails(String userName, String projectID) {
        if (Strings.isNullOrEmpty(userName) || Strings.isNullOrEmpty(projectID)) {
            return Observable.error(new IllegalArgumentException("Username and projectID must be provided!"));
        }

        return projectListRepository.getProjectDetails(userName, projectID)
                .subscribeOn(Schedulers.io());
    }
}

File: app/src/main/kotlin/com/test/xyz/demo/domain/interactor/project/ProjectInteractorImpl.kt
package com.test.xyz.demo.domain.interactor.project

import com.google.common.base.Strings
import com.test.xyz.demo.domain.model.github.GitHubRepo
import com.test.xyz.demo.domain.repository.api.ProjectListRepository

import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers

internal class ProjectInteractorImpl(private val projectListRepository: ProjectListRepository) : ProjectInteractor {

    override fun getProjectList(userName: String): Observable<List<GitHubRepo>> {
        return if (Strings.isNullOrEmpty(userName)) {
            Observable.error(IllegalArgumentException("Username must be provided!"))
        } else projectListRepository.getProjectList(userName)
                .subscribeOn(Schedulers.io())

    }

    override fun getProjectDetails(userName: String, projectID: String): Observable<GitHubRepo> {
        return if (Strings.isNullOrEmpty(userName) || Strings.isNullOrEmpty(projectID)) {
            Observable.error(IllegalArgumentException("Username and projectID must be provided!"))
        } else projectListRepository.getProjectDetails(userName, projectID)
                .subscribeOn(Schedulers.io())

    }

}
